e2e/actions/l2_challenger_test.go

```diff
@@ -197,3 +197,181 @@ interaction:
+
+	// deposit bond for validator
+	validator.ActDeposit(t, 1_000)
+	includeL1Block(t, miner, validator.address)
+
+	require.Equal(t, proposer.SyncStatus().UnsafeL2, proposer.SyncStatus().FinalizedL2)
+
+	outputOracleContract, err := bindings.NewL2OutputOracle(sd.DeploymentsL1.L2OutputOracleProxy, miner.EthClient())
+	require.NoError(t, err)
+
+	validatorRPC.SetTargetBlockNumber(testdata.TargetBlockNumber)
+	challengerRPC.SetTargetBlockNumber(testdata.TargetBlockNumber)
+	guardianRPC.SetTargetBlockNumber(testdata.TargetBlockNumber)
+	// create l2 output submission transactions until there is nothing left to submit
+	for validator.CanSubmit(t) {
+		// and submit it to L1
+		validator.ActSubmitL2Output(t)
+		// include output on L1
+		includeL1Block(t, miner, validator.address)
+		// Check submission was successful
+		receipt, err := miner.EthClient().TransactionReceipt(t.Ctx(), validator.LastSubmitL2OutputTx())
+		require.NoError(t, err)
+		require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status, "submission failed")
+	}
+
+	// check that L1 stored the expected output root
+	// NOTE(chokobole): Comment these 2 lines because of the reason above about the Blue hard fork.
+	// If Proto Dank Sharding is introduced, the below code fix may be restored.
+	// block := proposer.SyncStatus().FinalizedL2
+	// outputOnL1, err := outputOracleContract.GetL2OutputAfter(nil, new(big.Int).SetUint64(block.Number))
+	blockNum := big.NewInt(int64(testdata.TargetBlockNumber))
+	outputIndex, err := outputOracleContract.GetL2OutputIndexAfter(nil, blockNum)
+	require.NoError(t, err)
+	outputOnL1, err := outputOracleContract.GetL2OutputAfter(nil, blockNum)
+	require.NoError(t, err)
+	block, err := propEngine.EthClient().BlockByNumber(t.Ctx(), blockNum)
+	require.NoError(t, err)
+	require.Less(t, block.Time(), outputOnL1.Timestamp.Uint64(), "output is registered with L1 timestamp of L2 tx output submission, past L2 block")
+	outputComputed, err := challengerRollupClient.OutputAtBlock(t.Ctx(), blockNum.Uint64())
+	require.NoError(t, err)
+	require.NotEqual(t, eth.Bytes32(outputOnL1.OutputRoot), outputComputed.OutputRoot, "output roots must different")
+
+	// deposit bond for challenger
+	challenger.ActDeposit(t, 1_000)
+	includeL1Block(t, miner, challenger.address)
+
+	// submit create challenge tx
+	txHash := challenger.ActCreateChallenge(t, outputIndex)
+
+	// include tx on L1
+	includeL1Block(t, miner, challenger.address)
+
+	// Check whether the submission was successful
+	receipt, err := miner.EthClient().TransactionReceipt(t.Ctx(), txHash)
+	require.NoError(t, err)
+	require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status, "failed to create challenge")
+
+	// check challenge created
+	colosseumContract, err := bindings.NewColosseum(sd.DeploymentsL1.ColosseumProxy, miner.EthClient())
+	require.NoError(t, err)
+	challenge, err := colosseumContract.GetChallenge(nil, outputIndex)
+	require.NoError(t, err)
+	require.NotNil(t, challenge, "challenge not found")
```

Comment:
How about separating the repeated code blocks in this test and `TestChallenger` into another function? Quite many lines between L202-L337 are repeated except for guardian part.

Reply:
Good idea. I'll make it reusable.

Reply:
It will be refactored in https://github.com/kroma-network/kroma/pull/85. So I approve this PR without this change.